### PR TITLE
fix(ecs): EcsOptimizedImage and EcsOptimizedAmi are not using the correct SSM parameter names for windows images

### PR DESCRIPTION
**Pull request by 47star**
The [WindowsOptimizedVersion](https://github.com/aws/aws-cdk/blob/174b066634755c76d1b78d05ca9b403145dedc47/packages/%40aws-cdk/aws-ecs/lib/cluster.ts#L576) provides the oldest release, which is not same with the AMI that was defined as latest release from [Amazon ECS-optimized AMI versions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-windows-ami-versions.html). Also, It does not seem to be maintained anymore, and the Docker engine it contains is too old.

For this reasons, I changed the followings.
1. Mark `WindowsOptimizedVersion.SERVER_2016` and `SERVER_2019` as deprecated and change to provide latest ECS optimized Windows Full(not Core) AMI.
2. Add `SERVER_2016_FULL`, `SERVER_2016_CORE`, `SERVER_2019_FULL`, `SERVER_2019_CORE`, `SERVER_2004_CORE` and `SERVER_20H2_CORE` into `WindowsOptimizedVersion`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by donggu-kang** [comment]
I too have experienced the same issue, and this issue should be resolved soon.

**Comment by RexMorgan** [comment]
Any plans to get this merged? I wonder how many people are running outdated Windows ECS clusters out there because they didn't realize it was using an AMI that's almost 3 years old at this point.

If anyone else noticed their cluster was spinning up with the old Windows AMI, this is what I updated my cluster `machineImage` to in order to get the latest AMI.

```
machineImage: MachineImage.fromSsmParameter('/aws/service/ami-windows-latest/Windows_Server-2019-English-Core-ECS_Optimized/image_id', { os: OperatingSystemType.WINDOWS })
```

The full list of SSM parameters for the different versions of Windows is listed here.
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/retrieve-ecs-optimized_windows_AMI.html

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts [hunk 1]
```diff
@@ -724,13 +740,16 @@ export class EcsOptimizedImage implements ec2.IMachineImage {
     }
 
     // set the SSM parameter name
-    this.amiParameterName = '/aws/service/ecs/optimized-ami/'
+    if (this.windowsVersion) {
+      this.amiParameterName = `/aws/service/ami-windows-latest/${this.windowsVersion}-ECS_Optimized/image_id`;
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts:744]: This is a breaking change for users who already have an ssm parameter name in the format `/aws/service/ecs/optimized-ami/windows_server/<VERSION>/english/full` in their templates. So let's only use this format if the user supplies one of the new enum values defined in this PR.

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts [hunk 2]
```diff
@@ -574,8 +574,21 @@ export class Cluster extends Resource implements ICluster {
  * ECS-optimized Windows version list
  */
 export enum WindowsOptimizedVersion {
-  SERVER_2019 = '2019',
-  SERVER_2016 = '2016',
+  SERVER_20H2_CORE = 'Windows_Server-20H2-English-Core',
+  SERVER_2004_CORE = 'Windows_Server-2004-English-Core',
+  SERVER_2019_FULL = 'Windows_Server-2019-English-Full',
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts:579]: We can't use enums where there are multiple different enum entries with the same string values (this one and `SERVER_2016_FULL`). This unfortunately causes unexpected behavior when using the CDK in non-typescript languages. 

However, this kind of fixes itself if you revert the `SERVER_2019` and `SERVER_2016` entries back to their old values to not break existing users.

**Review by madeline-k — CHANGES_REQUESTED** [review]
Thank you for making this contribution, @donghoon-yoo! And thanks for your patience, it took us a little while to get to this one. 

The main feedback I have is that we need to keep the old logic for the deprecated WindowsOptimizedVersion enum values. And only use the new SSM parameter values with the new WindowsOptimizedVersion enum values. 

I've also updated the Pull Request description to remove the breaking change section, since we cannot add any breaking changes to stable modules. That's fine though, since we can implement this fix without a breaking change.

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts (hunk 2 above)
Comment (RexMorgan) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts:579]: Hi @madeline-k, this isn't my PR, however the string values for `SERVER_2016_FULL` and `SERVER_2019_FULL` have different strings, one for 2016 and one for 2019, but I could be misunderstanding the issue here.

Comment (madeline-k) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts:579]: Sorry, my comment wasn't clear. the string for SERVER_2019_FULL is the same as the string for SERVER_2019. And SERVER_2016_FULL is the same as SERVER_2016.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be closed. Please create a new PR for these changes if you think this decision has been made in error.

**Pull request closed without merging**